TAO/TAO_IDL/ast/ast_map.cpp

```diff
@@ -0,0 +1,374 @@
+              n,
+              true),
+    AST_Type (AST_Decl::NT_map,
+              n),
+    AST_ConcreteType (AST_Decl::NT_map,
+                      n),
+    pd_max_size (ms),
+    key_pd_type (key_bt),
+    value_pd_type (val_bt),
+    unbounded_ (true),
+    owns_key_type_ (false),
+    owns_value_type_ (false)
+{
+  FE_Utils::tmpl_mod_ref_check (this, key_bt);
+
+  AST_Decl::NodeType knt = key_bt->node_type ();
+
+  if (knt == AST_Decl::NT_param_holder)
+    {
+      AST_Param_Holder *ph = dynamic_cast<AST_Param_Holder *> (key_bt);
+
+      if (ph->info ()->type_ == AST_Decl::NT_const)
+        {
+          idl_global->err ()->not_a_type (key_bt);
+          key_bt->destroy ();
+          delete key_bt;
+          key_bt = nullptr;
+          throw Bailout ();
+        }
+    }
+
+  FE_Utils::tmpl_mod_ref_check (this, val_bt);
+
+  AST_Decl::NodeType vnt = val_bt->node_type ();
+
+  if (vnt == AST_Decl::NT_param_holder)
+    {
+      AST_Param_Holder *ph = dynamic_cast<AST_Param_Holder*> (val_bt);
+
+      if (ph->info ()->type_ == AST_Decl::NT_const)
+        {
+          idl_global->err ()->not_a_type (val_bt);
+          val_bt->destroy ();
+          delete val_bt;
+          val_bt = nullptr;
+          throw Bailout ();
+        }
+    }
+
+  // Check if we are bounded or unbounded. An expression value of 0 means
+  // unbounded. If our bound is a template parameter, skip the
+  // check altogether, this node will trigger no code generation.
+  if (ms->param_holder () == nullptr)
+    {
+      this->unbounded_ = (ms->ev ()->u.ulval == 0);
+    }
+
+  // A map data type is always VARIABLE.
+  this->size_type (AST_Type::VARIABLE);
+
+  this->owns_key_type_ =
+    knt == AST_Decl::NT_array
+    || knt == AST_Decl::NT_map
+    || knt == AST_Decl::NT_param_holder;
+
+  this->owns_value_type_ =
+    vnt == AST_Decl::NT_array
+    || vnt == AST_Decl::NT_map
+    || vnt == AST_Decl::NT_param_holder;
+}
+
+// Public operations.
+
+bool
+AST_Map::in_recursion (ACE_Unbounded_Queue<AST_Type *> &list)
+{
+  if (list.is_empty ()) // only structs, unions and valuetypes can be recursive
+    return false;
+
+  list.enqueue_tail(this);
+
+  AST_Type *type = dynamic_cast<AST_Type*> (this->key_type ());
+
+  if (type == nullptr)
+    {
+      ACE_ERROR_RETURN ((LM_ERROR,
+                         ACE_TEXT ("AST_Map::in_recursion - ")
+                         ACE_TEXT ("bad base type\n")),
+                        false);
+    }
+
+  AST_Decl::NodeType nt = type->node_type ();
+
+  if (nt == AST_Decl::NT_typedef)
+    {
+      AST_Typedef *td = dynamic_cast<AST_Typedef*> (type);
+      type = td->primitive_base_type ();
+      nt = type->node_type ();
+    }
+
+  if (nt != AST_Decl::NT_struct
+      && nt != AST_Decl::NT_union
+      && nt != AST_Decl::NT_valuetype
+      && nt != AST_Decl::NT_map)
+    {
+      return false;
+    }
+
+  bool recursion_found = false;
+  AST_Type** recursable_type = nullptr;
+  list.get (recursable_type, 0);
```

Comment:
Can this fail, is there a case recursable_type would stay null?

Reply:
I'm not sure about this. I will admit most of this PR was copying the sequence's implementation and putting out fires along the way and back-filling the understanding as I went. So with that I'm not sure what makes a type recursable.

Reply:
Sequences can be declared recursive, can the same be done with maps? If so, please test that, if not, cleanup the recursive support for maps.

```
struct Foo; // Forward declaration
typedef sequence<Foo> FooSeq;
struct Foo {
long value;
FooSeq chain; // Recursive
}
```

Reply:
What does the IDL spec say, does it specify maps can be recursive, if so, we should support it

Reply:
[We discussed this earlier](https://github.com/DOCGroup/ACE_TAO/pull/1842#discussion_r888481661). I think what we agreed on is the spec doesn't say it explicitly, but it basically says maps are basically the same as sequences with associative behavior, so the same recursive rule should apply. Adam [filed a spec issue](https://issues.omg.org/issues/task-force/IDL43#issue-49295) for this to be explicit.

Reply:
This doesn't seem to fail, on either both recursed maps and normal maps (added a test for this). Althought the fuction was previously incorrect, should be good now